core/src/main/java/com/linecorp/armeria/server/AbstractHttpResponseHandler.java

```diff
@@ -0,0 +1,210 @@
+
+    /**
+     * Fails a request and a response with the specified {@link Throwable}.
+     * This method won't send any response to the remote peer but log the failed request.
+     */
+    abstract void fail(Throwable cause);
+
+    final boolean failIfStreamOrSessionClosed() {
+        // Make sure that a stream exists before writing data.
+        // The following situation may cause the data to be written to a closed stream.
+        // 1. A connection that has pending outbound buffers receives GOAWAY frame.
+        // 2. AbstractHttp2ConnectionHandler.close() clears and flushes all active streams.
+        // 3. After successfully flushing, the listener requests next data and
+        //    the subscriber attempts to write the next data to the stream closed at 2).
+        if (!isWritable()) {
+            Throwable cause = CapturedServiceException.get(reqCtx);
+            if (cause == null) {
+                if (reqCtx.sessionProtocol().isMultiplex()) {
+                    cause = ClosedSessionException.get();
+                } else {
+                    cause = ClosedStreamException.get();
+                }
+            }
+            fail(cause);
+            return true;
+        }
+        return false;
+    }
+
+    final boolean isWritable() {
+        return responseEncoder.isWritable(req.id(), req.streamId());
+    }
+
+    /**
+     * Writes the {@link AggregatedHttpResponse} to the {@link Channel}.
+     * Note that the caller has to flush the written data when needed.
+     */
+    final ChannelFuture writeAggregatedHttpResponse(AggregatedHttpResponse res) {
+        final int id = req.id();
+        final int streamId = req.streamId();
+
+        final ResponseHeaders headers = mergeResponseHeaders(res.headers(), reqCtx.additionalResponseHeaders());
+        final HttpData content = res.content();
+        final boolean contentEmpty = content.isEmpty();
+        final HttpHeaders trailers = mergeTrailers(res.trailers(), reqCtx.additionalResponseTrailers());
+        final boolean trailersEmpty = trailers.isEmpty();
+
+        if (!res.informationals().isEmpty()) {
+            for (ResponseHeaders informational : res.informationals()) {
+                responseEncoder.writeHeaders(id, streamId, informational,
+                                             false, trailersEmpty);
+            }
+        }
+        logBuilder().responseHeaders(headers);
+        ChannelFuture future = responseEncoder.writeHeaders(id, streamId, headers,
+                                                            contentEmpty && trailersEmpty, trailersEmpty);
+        if (!contentEmpty) {
+            logBuilder().increaseResponseLength(content);
+            future = responseEncoder.writeData(id, streamId, content, trailersEmpty);
+        }
+        if (!trailersEmpty) {
+            logBuilder().responseTrailers(trailers);
+            future = responseEncoder.writeTrailers(id, streamId, trailers);
+        }
```

Comment:
Can we set `content-length` when we're not chunked encoding? i.e.
- If HTTP/2, always set `content-length`
- If HTTP/1 and trailers is empty, set `content-length`

Reply:
and some test cases for this. But perhaps this could be addressed in a follow-up PR?

Reply:
Good suggestion. It seems that we can reuse the following method.
https://github.com/line/armeria/blob/825af6352c440187607b3d3f5dc87ec861922ef6/core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java#L974
Let me address this in this PR.